compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp

```diff
@@ -21,12 +22,54 @@
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 #include "mlir/Transforms/Passes.h"
 
+#define DEBUG_TYPE "iree-codegen-gpu-tensorcore-vectorization"
+
 using mlir::iree_compiler::IREE::LinalgExt::LinalgVectorizationPattern;
 using mlir::iree_compiler::IREE::LinalgExt::VectorizationPatterns;
 
 namespace mlir {
 namespace iree_compiler {
 
+/// Pattern to fold arithmetic extensions on floating point data types into
+/// vector contraction operations. linalg.matmul introduces arithmetic
+/// extensions on its operands. Please mlir snippets below for more details.
+/// ```mlir
+///   "linalg.matmul"(%lhs, %rhs, %acc) ({
+///      ^bb0(%arg1: f16, %arg2: f16, %arg3: f32):
+///        %lhs_f32 = "arith.extf"(%arg1) : (f16) -> f32
+///        %rhs_f32 = "arith.extf"(%arg2) : (f16) -> f32
+///        %mul = "arith.mulf"(%lhs_f32, %rhs_f32) : (f32, f32) -> f32
+///        %acc = "arith.addf"(%arg3, %mul) : (f32, f32) -> f32
+///        "linalg.yield"(%acc) : (f32) -> ()
+///     })
+/// ```
+/// This restricts the native usage of mixed precision NVIDIA Ampere Tensor
+/// Cores, i.e, `mma.sync.*.f32.f16.f16.f32` and `mma.sync.*.f32.bf16.bf16.f32`.
+/// This pattern folds the arithmetic extensions into the vector contraction and
+/// enables the usage of native mixed precision Tensor Core instructions.
+struct FoldArithExtIntoContractionOp
```

Comment:
Maybe add this transformation upstream... 
Ideally we could have a way to check if a type is legal or not for a given architecture and then fold it. Not saying this is a hard requirement though, but adding this upstream is better overall.

Reply:
+1 on upstreaming.
Also this is missing a test case.

Reply:
my 2cents about upstreaming - while upstreaming offers numerous benefits, it can be time-consuming. If this transformation won't be utilized by any other piece, we can delay upstreaming to speed up development.

Reply:
I have moved the folding pattern upstream and added a test, but the test is failing when run stand-alone, can someone please check, what i did incorrectly. LLVM upstream diff [here](https://github.com/iree-org/iree-llvm-fork/compare/fold_arith_ext_into_vector_contract?expand=1)

PS: I will create Phabricator patch once I get it working.

Reply:
What is the error?

Reply:
If that's too hard to make it work on upstream LLVM, write a test with `iree-opt` for now.

Reply:
In my experience upstreaming never happens. So I tend to push for adding this upstream first, and only considering putting this in iREE when it is not possible to add upstream. This is fairly hermitic, but just as a process point/convention would prefer this upstream.

Reply:
Please review the upstream patch [D151918](https://reviews.llvm.org/D151918). I think this is very NVIDIA A100/H100 specific folding. 

The developmental steps to now get to #13858 (F32 <= BF16 * BF16 + F32) from #13857 (F32 <= F16 * F16  + F32): 
- Land patch [D151918](https://reviews.llvm.org/D151918) 
- Have iree llvm-integrate pick up the patch 
- Merge this PR #13857 (F32 <= F16 * F16  + F32)
- Rebase #13858 (F32 <= BF16 * BF16 + F32) 
- Proceed with #13858 (F32 <= BF16 * BF16 + F32).

Reply:
@all Does the patch D151918 look ok to land? 

Then we wait on llvm-integrate to pick this up and land this PR.

Reply:
Land patch [D151918](https://reviews.llvm.org/D151918) has landed. Now we are going to wait for the patch to be picked up by llvm-integrate before we land this PR.

Reply:
We are now ready to merge this PR, [D151918](https://reviews.llvm.org/D151918) has landed, and we have rebased this PR. This PR now only contains IREE chagnes and no upstream changes